src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -188,5 +256,159 @@ private bool AssemblyContainsTagHelpers(string assemblyName, IReadOnlyList<TagHe
+                    }
+                    else if (context.ChunkGenerator is AddTagHelperChunkGenerator addTagHelper)
+                    {
+                        // Make sure this node exists in the file we're parsing and not in its imports.
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
+                        {
+                            addTagHelper.Diagnostics.Add(
+                                ComponentDiagnosticFactory.Create_UnsupportedTagHelperDirective(node.GetSourceSpan(_source)));
+                        }
+                    }
+                    else if (context.ChunkGenerator is RemoveTagHelperChunkGenerator removeTagHelper)
+                    {
+                        // Make sure this node exists in the file we're parsing and not in its imports.
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
+                        {
+                            removeTagHelper.Diagnostics.Add(
+                                ComponentDiagnosticFactory.Create_UnsupportedTagHelperDirective(node.GetSourceSpan(_source)));
+                        }
+                    }
+                    else if (context.ChunkGenerator is TagHelperPrefixDirectiveChunkGenerator tagHelperPrefix)
+                    {
+                        // Make sure this node exists in the file we're parsing and not in its imports.
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
+                        {
+                            tagHelperPrefix.Diagnostics.Add(
+                                ComponentDiagnosticFactory.Create_UnsupportedTagHelperDirective(node.GetSourceSpan(_source)));
+                        }
+                    }
+                    else if (context.ChunkGenerator is AddImportChunkGenerator usingStatement && !usingStatement.IsStatic)
+                    {
+                        // Get the namespace from the using statement. Split it at '=' in case there is an alias.
+                        var @namespace = usingStatement.ParsedNamespace.Split('=').LastOrDefault();
+                        for (var i = 0; i < _tagHelpers.Count; i++)
+                        {
+                            var tagHelper = _tagHelpers[i];
+                            if (tagHelper.IsComponentFullyQualifiedNameMatch())
+                            {
+                                // We've already added these to our list of matches.
+                                continue;
+                            }
+
+                            var typeName = tagHelper.GetTypeName();
+                            if (typeName != null && IsTypeInNamespace(typeName, @namespace))
+                            {
+                                // If the type is at the top-level or if the type's namespace matches the using's namespace, add it.
+                                Matches.Add(tagHelper);
+                            }
```

Comment:
Do we report ambiguities?

Reply:
You might want to think about separating this logic from the visitor into something that's more directly unit testable.

Step 1: Accumulate namespaces that are in scope
Step 2: Build a lookup table of name -> tag helpers based on data from 1
Step 3: Iterate matches and use the lookup table

This is the kind of logic I expected to see here, and it seems like it would be necessary to know about ambiguities.

What's the rule about what happens when an `@using` brings in something that overlaps with the document's current namespace? Put another way, what's the difference between putting usings outside the namespace declaration vs inside (hint, there is a semantic difference in C#)?

Reply:
Spoke offline. My plan was to let the C# compiler report ambiguities now that we generate design time code gen for each of the component we use. I'll probably need to add line mappings to it though. I'll switch to using a lookup table if that turns out to be complicated/not what we want.